include/alpaka/warp/WarpGenericSycl.hpp

```diff
@@ -132,6 +131,68 @@ namespace alpaka::warp::trait
             return sycl::select_from_group(actual_group, value, src);
         }
     };
+
+    template<typename TDim>
+    struct ShflUp<warp::WarpGenericSycl<TDim>>
+    {
+        template<typename T>
+        static auto shfl_up(
+            warp::WarpGenericSycl<TDim> const& warp,
+            T value,
+            std::uint32_t offset, /* must be the same for all work-items in the group */
+            std::int32_t width)
+        {
+            std::int32_t offset_int = static_cast<std::int32_t>(offset);
+            auto const actual_group = warp.m_item_warp.get_sub_group();
+            auto actual_item_id = static_cast<std::int32_t>(actual_group.get_local_linear_id());
+            auto const actual_group_id = actual_item_id / width;
+            auto const actual_src_id = actual_item_id - offset_int;
+            auto const src = actual_src_id >= actual_group_id * width
+                                 ? sycl::id<1>{static_cast<std::size_t>(actual_src_id)}
+                                 : sycl::id<1>{static_cast<std::size_t>(actual_item_id)};
+            return sycl::select_from_group(actual_group, value, src);
+        }
+    };
+
+    template<typename TDim>
+    struct ShflDown<warp::WarpGenericSycl<TDim>>
+    {
+        template<typename T>
+        static auto shfl_down(
+            warp::WarpGenericSycl<TDim> const& warp,
+            T value,
+            std::uint32_t offset,
+            std::int32_t width)
+        {
+            std::int32_t offset_int = static_cast<std::int32_t>(offset);
+            auto const actual_group = warp.m_item_warp.get_sub_group();
+            auto actual_item_id = static_cast<std::int32_t>(actual_group.get_local_linear_id());
+            auto const actual_group_id = actual_item_id / width;
+            auto const actual_src_id = actual_item_id + offset_int;
+            auto const src = actual_src_id < (actual_group_id + 1) * width
+                                 ? sycl::id<1>{static_cast<std::size_t>(actual_src_id)}
+                                 : sycl::id<1>{static_cast<std::size_t>(actual_item_id)};
+            return sycl::select_from_group(actual_group, value, src);
+        }
+    };
+
+    template<typename TDim>
+    struct ShflXor<warp::WarpGenericSycl<TDim>>
+    {
+        template<typename T>
+        static auto shfl_xor(
+            warp::WarpGenericSycl<TDim> const& warp,
+            T value,
+            std::int32_t mask,
+            std::int32_t /*width*/)
```

Comment:
The CUDA version does make use of width:

> `__shfl_xor_sync()` calculates a source line ID by performing a bitwise XOR of the caller’s lane ID with `laneMask`: the value of `var` held by the resulting lane ID is returned. If `width` is less than `warpSize` then each group of `width` consecutive threads are able to access elements from earlier groups of threads, however if they attempt to access elements from later groups of threads their own value of `var` will be returned. This mode implements a butterfly addressing pattern such as is used in tree reduction and broadcast.

Can you implement the same behaviour ?